### PR TITLE
[lib] Add `Map.update` from OCaml 4.06

### PR DESCRIPTION
**Pull request by ejgallego**
It will take more than a year to bump the OCaml version, this is in
response of a request by @Skyskimmer.

We also update our internal repr to make it closer to the one in
modern OCaml.

### CODE REVIEW

File: clib/hMap.ml [hunk 1]
```diff
@@ -408,6 +408,19 @@ struct
   let filter_range f s =
     filter (fun x _ -> f x = 0) s
 
+  (* Not as efficient as the original version *)
```
Comment (SkySkimmer) [inline clib/hMap.ml]: Pushed a hopefully more efficient version.

**Review by SkySkimmer — APPROVED** [review]
It's a nice function

File: clib/hMap.ml (hunk 1 above)
Comment (ejgallego) [reply clib/hMap.ml]: Merci

**Comment by ejgallego** [comment]
Ping @ppedrot , alternatively @SkySkimmer if this is useful for you IMHO [as secondary] you can assign to you and merge.

**Review by ppedrot — APPROVED** [review]

**Comment by ejgallego** [comment]
Merci !

**Pull request merged**